# Data Wrangling exercise

In [1]:
#Importing all required modules
import pandas as pd
import numpy as np
import json

In [2]:
# load world bank project dataset as Pandas dataframe
wbp_df = pd.read_json('data/world_bank_projects.json')
wbp_df.head()


_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

   ...   sectorcode source  status  supplementprojectflg  \
0  ...  ET,BS,ES,EP   IBRD  Active                     N   
1  ...        BZ,BS   IBRD  Active                     N   
2  ...           TI   IBRD  Active                     Y   
3  ...           JB   IBRD  Active                     N   
4  ...     FH,YW,YZ   IBRD  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [3]:
wbp_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

In [4]:
wbp_df.shape

(500, 50)

### Extract only required columns for the excercise into new DataFrame.

In [5]:
projects_df=wbp_df.loc[:,['countryname','country_namecode','countrycode','project_name','mjtheme_namecode']]
projects_df.head()

countryname  \
0  Federal Democratic Republic of Ethiopia   
1                      Republic of Tunisia   
2                                   Tuvalu   
3                        Republic of Yemen   
4                       Kingdom of Lesotho   

                               country_namecode countrycode  \
0  Federal Democratic Republic of Ethiopia!$!ET          ET   
1                      Republic of Tunisia!$!TN          TN   
2                                   Tuvalu!$!TV          TV   
3                        Republic of Yemen!$!RY          RY   
4                       Kingdom of Lesotho!$!LS          LS   

                                        project_name  \
0  Ethiopia General Education Quality Improvement...   
1          TN: DTF Social Protection Reforms Support   
2  Tuvalu Aviation Investment Project - Additiona...   
3   Gov't and Civil Society Organization Partnership   
4  Second Private Sector Competitiveness and Econ...   

                                    mjtheme_namecode  
0  [{'code': '8', 'name': 'Human development'}, {...  
1  [{'code': '1', 'name': 'Economic management'},...  
2  [{'code': '5', 'name': 'Trade and integration'...  
3  [{'code': '7', 'name': 'Social dev/gender/incl...  
4  [{'code': '5', 'name': 'Trade and integration'...

## Excercise 1 : Top 10 countries with most projects - Approach 1

In [6]:
#Grouping the entire Datframe by Country to get the count of Projects
groupedProjects=projects_df.groupby('countryname')['project_name'].count().reset_index()
#Renaming the column names to meaningful one's 
groupedProjects.columns = ['Country Name','Project Count']
#Sorting in descending order of Project counts in order filter top 10
sortedGroupedProjects = groupedProjects.sort_values('Project Count', ascending=False)
print(sortedGroupedProjects.head(10))


                        Country Name  Project Count
39        People's Republic of China             19
64             Republic of Indonesia             19
107    Socialist Republic of Vietnam             17
63                 Republic of India             16
97                 Republic of Yemen             13
38   People's Republic of Bangladesh             12
34                             Nepal             12
25                Kingdom of Morocco             12
76            Republic of Mozambique             11
0                             Africa             11


### Approach 2 : For finding Top 10 countries with most projects.

In [7]:
projects_df.countryname.value_counts().head(10)

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Kingdom of Morocco                 12
People's Republic of Bangladesh    12
Nepal                              12
Africa                             11
Republic of Mozambique             11
Name: countryname, dtype: int64

### Random thought to see all projects done in India

In [8]:
is_india_projects =  projects_df['countryname']=='Republic of India'
india_projects = projects_df[is_india_projects]

In [9]:
india_projects

countryname        country_namecode countrycode  \
6    Republic of India  Republic of India!$!IN          IN   
8    Republic of India  Republic of India!$!IN          IN   
11   Republic of India  Republic of India!$!IN          IN   
57   Republic of India  Republic of India!$!IN          IN   
122  Republic of India  Republic of India!$!IN          IN   
195  Republic of India  Republic of India!$!IN          IN   
196  Republic of India  Republic of India!$!IN          IN   
214  Republic of India  Republic of India!$!IN          IN   
394  Republic of India  Republic of India!$!IN          IN   
431  Republic of India  Republic of India!$!IN          IN   
452  Republic of India  Republic of India!$!IN          IN   
456  Republic of India  Republic of India!$!IN          IN   
458  Republic of India  Republic of India!$!IN          IN   
481  Republic of India  Republic of India!$!IN          IN   
482  Republic of India  Republic of India!$!IN          IN   
483  Republic of India  Republic of India!$!IN          IN   

                                          project_name  \
6    National Highways Interconnectivity Improvemen...   
8          Rajasthan Road Sector Modernization Project   
11               Uttarakhand Disaster Recovery Project   
57   Uttar Pradesh Water Sector Restructuring Proje...   
122  Tamil Nadu and Puducherry Coastal Disaster Ris...   
195                   India Low-Income Housing Finance   
196        India Second Kerala State Transport Project   
214              National AIDS Control Support Project   
394      Himachal Pradesh Watershed Management Project   
431      HP State Roads Project - Additional Financing   
452      India - Bihar Panchayat Strengthening Project   
456  India:  Karnataka Health Systems Additional Fi...   
458  AF - HP Mid-Himalayan Watershed Development Pr...   
481  India: ICDS Systems Strengthening & Nutrition ...   
482                 Karnataka Watershed Development II   
483  IN: Development Policy Loan (DPL) to Promote I...   

                                      mjtheme_namecode  
6    [{'code': '2', 'name': 'Public sector governan...  
8    [{'code': '10', 'name': 'Rural development'}, ...  
11   [{'code': '10', 'name': 'Rural development'}, ...  
57   [{'code': '11', 'name': 'Environment and natur...  
122  [{'code': '6', 'name': 'Social protection and ...  
195  [{'code': '9', 'name': 'Urban development'}, {...  
196  [{'code': '4', 'name': 'Financial and private ...  
214  [{'code': '8', 'name': 'Human development'}, {...  
394  [{'code': '11', 'name': 'Environment and natur...  
431  [{'code': '4', 'name': 'Financial and private ...  
452  [{'code': '10', 'name': 'Rural development'}, ...  
456  [{'code': '6', 'name': 'Social protection and ...  
458  [{'code': '10', 'name': 'Rural development'}, ...  
481  [{'code': '8', 'name': 'Human development'}, {...  
482  [{'code': '10', 'name': 'Rural development'}, ...  
483  [{'code': '11', 'name': 'Environment and natur...

## Excercise 2 : Find the top 10 major project themes (using column 'mjtheme_namecode')  : Start 

In [10]:
#Fetching data from mjtheme_namecode column of a random row (65th Row here) to have an understanding of structure and contents.
df1=pd.DataFrame(projects_df['mjtheme_namecode'][65])
df1

code                                      name
0    4  Financial and private sector development
1    7               Social dev/gender/inclusion
2   10                         Rural development
3    5                     Trade and integration

In [11]:
# Creating and appending to a new Dataframe to store all project theme's in one location
projectTheme_df=pd.DataFrame() 
for index, row in projects_df.iterrows():
    df1=pd.DataFrame(row['mjtheme_namecode'])
    projectTheme_df = projectTheme_df.append(df1)
      
projectTheme_df = projectTheme_df.reset_index()
projectTheme_df = projectTheme_df.drop(['index'], axis = 1)
projectTheme_df.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [12]:
projectTheme_df.shape

(1499, 2)

### List of Top 10 Project theme's from the entire World Bank Projects Dataset

In [13]:
projectTheme_df.name.value_counts().head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

## Excercise 3: Fill all the columns with empty project theme by the correct one(Total 122 row's have empty project name cells)

## Approach 1 : Sorting by code,name cells combination and using backward filling technique 

In [14]:
# Grouping the code and name field combo to get the count of combinations
uniqueValues = projectTheme_df.groupby(['code','name']).size().reset_index().rename(columns={0:'count'})

# Changing the type of code field in the datframe to integer for effective sorting
uniqueValues = uniqueValues.astype({'code': 'int32'})
uniqueValues = uniqueValues.sort_values('code', ascending=True).reset_index(drop=True)
print(uniqueValues)

    code                                          name  count
0      1                                                    5
1      1                           Economic management     33
2      2                      Public sector governance    184
3      2                                                   15
4      3                                   Rule of law     12
5      3                                                    3
6      4      Financial and private sector development    130
7      4                                                   16
8      5                                                    5
9      5                         Trade and integration     72
10     6                                                   10
11     6         Social protection and risk management    158
12     7                                                   11
13     7                   Social dev/gender/inclusion    119
14     8                                                   13
15     8

In [15]:
# Droping all rows with empty name cells to get all unique name combinations alone for each code.
uniqueValues.drop(uniqueValues[uniqueValues['name'] == ''].index, inplace = True)
uniqueValues = uniqueValues.reset_index(drop=True)
#uniqueValues = uniqueValues[uniqueValues['name'].map(lambda x: str(x)=="")]
#uniqueValues[(len(uniqueValues['code']) < 0)]
uniqueValues

code                                          name  count
0      1                           Economic management     33
1      2                      Public sector governance    184
2      3                                   Rule of law     12
3      4      Financial and private sector development    130
4      5                         Trade and integration     72
5      6         Social protection and risk management    158
6      7                   Social dev/gender/inclusion    119
7      8                             Human development    197
8      9                             Urban development     47
9     10                             Rural development    202
10    11  Environment and natural resources management    223

***
+ From the initial lookup into the project theme dataframe it is understood that there are 2 fields relating to Project theme 
  in the dataframe, one is code and other project name. 
+ Also the Project Name and Code combination is unique. 
+ Since code and name combination is unique, we can replace the empty cells by checking the name corresponding to the same 
  code in other row's.
***

In [16]:
# Changing the type of code field in the datframe to integer for effective sorting
projectTheme_df = projectTheme_df.astype({'code': 'int32'})
projectTheme_df = projectTheme_df.sort_values(['code', 'name'], ascending=True).reset_index(drop=True)
projectTheme_df.head(10)

code                 name
0     1                     
1     1                     
2     1                     
3     1                     
4     1                     
5     1  Economic management
6     1  Economic management
7     1  Economic management
8     1  Economic management
9     1  Economic management

In [17]:
#projectTheme_df['name'] = projectTheme_df['name'].apply(lambda row:'' if pd.isna(row) else row)
#projectTheme_df['name'] = projectTheme_df['name'].apply(lambda row:np.nan if row=='' else row)

# Replacing all empty cells with NAN values as it will be easy to fill these values. 
projectTheme_df['name'] = projectTheme_df['name'].replace('',np.nan)
projectTheme_df.head(10)

code                 name
0     1                  NaN
1     1                  NaN
2     1                  NaN
3     1                  NaN
4     1                  NaN
5     1  Economic management
6     1  Economic management
7     1  Economic management
8     1  Economic management
9     1  Economic management

### Filling the NaN values in the project name cell with backward column filling technique

In [18]:
projectTheme_df["name"].fillna( method ='bfill', inplace = True)
projectTheme_df.head(15)

code                 name
0      1  Economic management
1      1  Economic management
2      1  Economic management
3      1  Economic management
4      1  Economic management
5      1  Economic management
6      1  Economic management
7      1  Economic management
8      1  Economic management
9      1  Economic management
10     1  Economic management
11     1  Economic management
12     1  Economic management
13     1  Economic management
14     1  Economic management

### Verifying the Final results to make sure that all empty cells have been replaced with correct project name values 

In [19]:
projectTheme_df.name.value_counts().head(15)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64

## Approach 2 : For filling all the columns with empty project theme by the correct one(By Merging Datframes) 

In [20]:
projectTheme_df2=pd.DataFrame() 
for index, row in projects_df.iterrows():
    df1=pd.DataFrame(row['mjtheme_namecode'])
    projectTheme_df2 = projectTheme_df2.append(df1)
      
projectTheme_df2 = projectTheme_df2.reset_index()
projectTheme_df2 = projectTheme_df2.drop(['index'], axis = 1)
projectTheme_df2.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [21]:
# Grouping the code and name field combo to get the count of combinations
uniqueValues = projectTheme_df2.groupby(['code','name']).size().reset_index().rename(columns={0:'count'})

# Changing the type of code field in the datframe to integer for effective sorting
uniqueValues = uniqueValues.astype({'code': 'int32'})
uniqueValues = uniqueValues.sort_values('code', ascending=True).reset_index(drop=True)
uniqueValues.drop(uniqueValues[uniqueValues['name'] == ''].index, inplace = True)
uniqueValues = uniqueValues.reset_index(drop=True)
uniqueValues

code                                          name  count
0      1                           Economic management     33
1      2                      Public sector governance    184
2      3                                   Rule of law     12
3      4      Financial and private sector development    130
4      5                         Trade and integration     72
5      6         Social protection and risk management    158
6      7                   Social dev/gender/inclusion    119
7      8                             Human development    197
8      9                             Urban development     47
9     10                             Rural development    202
10    11  Environment and natural resources management    223

In [22]:
projectTheme_df = projectTheme_df.astype({'code': 'int32'})
projectTheme_df = projectTheme_df.sort_values('code', ascending=True).reset_index(drop=True)

In [23]:
merged_df = pd.merge(projectTheme_df,uniqueValues,on="code")

In [24]:
merged_df.head(10)

code               name_x               name_y  count
0     1  Economic management  Economic management     33
1     1  Economic management  Economic management     33
2     1  Economic management  Economic management     33
3     1  Economic management  Economic management     33
4     1  Economic management  Economic management     33
5     1  Economic management  Economic management     33
6     1  Economic management  Economic management     33
7     1  Economic management  Economic management     33
8     1  Economic management  Economic management     33
9     1  Economic management  Economic management     33

In [25]:
merged_df.drop(['name_x', 'count'], axis=1,inplace=True)
merged_df.rename(columns={'name_y':'name'}, inplace=True)
merged_df.head(10)

code                 name
0     1  Economic management
1     1  Economic management
2     1  Economic management
3     1  Economic management
4     1  Economic management
5     1  Economic management
6     1  Economic management
7     1  Economic management
8     1  Economic management
9     1  Economic management

In [26]:
merged_df.name.value_counts().head(15)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64

## Approach 3 : To correct missing entries in dataframe(Long approach and not ideal one)

In [27]:
def printSelectedCodeFromDF(df,code):
    is_code = df['code']==code
    code_df = projectTheme_df[is_code]
    return code_df

code_df = printSelectedCodeFromDF(projectTheme_df,11)    
code_df.head(8)

code                                          name
1249    11  Environment and natural resources management
1250    11  Environment and natural resources management
1251    11  Environment and natural resources management
1252    11  Environment and natural resources management
1253    11  Environment and natural resources management
1254    11  Environment and natural resources management
1255    11  Environment and natural resources management
1256    11  Environment and natural resources management

In [28]:
def correct_missing(df,code,value):
    for index, row in df.iterrows():
        if (row['code'] == code and row['name']==''):
            df.iloc[index][1]=value
    
    return df    

In [29]:
projectTheme_df = correct_missing(projectTheme_df,11,'Environment and natural resources management')

In [30]:
code_df = printSelectedCodeFromDF(projectTheme_df,11)    
code_df.head(8)

code                                          name
1249    11  Environment and natural resources management
1250    11  Environment and natural resources management
1251    11  Environment and natural resources management
1252    11  Environment and natural resources management
1253    11  Environment and natural resources management
1254    11  Environment and natural resources management
1255    11  Environment and natural resources management
1256    11  Environment and natural resources management

In [31]:
projectTheme_df['code'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int64)

In [32]:
code_df = printSelectedCodeFromDF(projectTheme_df,8)    
code_df.head(6)

code               name
773     8  Human development
774     8  Human development
775     8  Human development
776     8  Human development
777     8  Human development
778     8  Human development

In [33]:
projectTheme_df = correct_missing(projectTheme_df,8,'Human development')

In [34]:
code_df = printSelectedCodeFromDF(projectTheme_df,1)    
code_df.head(10)

code                 name
0     1  Economic management
1     1  Economic management
2     1  Economic management
3     1  Economic management
4     1  Economic management
5     1  Economic management
6     1  Economic management
7     1  Economic management
8     1  Economic management
9     1  Economic management

In [35]:
projectTheme_df = correct_missing(projectTheme_df,1,'Economic management')

In [36]:
code_df = printSelectedCodeFromDF(projectTheme_df,6)    
code_df.head(8)

code                                   name
475     6  Social protection and risk management
476     6  Social protection and risk management
477     6  Social protection and risk management
478     6  Social protection and risk management
479     6  Social protection and risk management
480     6  Social protection and risk management
481     6  Social protection and risk management
482     6  Social protection and risk management

In [37]:
projectTheme_df = correct_missing(projectTheme_df,6,'Social protection and risk management')

In [38]:
code_df = printSelectedCodeFromDF(projectTheme_df,5)    
code_df.head(12)

code                   name
398     5  Trade and integration
399     5  Trade and integration
400     5  Trade and integration
401     5  Trade and integration
402     5  Trade and integration
403     5  Trade and integration
404     5  Trade and integration
405     5  Trade and integration
406     5  Trade and integration
407     5  Trade and integration
408     5  Trade and integration
409     5  Trade and integration

In [39]:
projectTheme_df = correct_missing(projectTheme_df,5,'Trade and integration')

In [40]:
projectTheme_df['code'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int64)

In [41]:
code_df = printSelectedCodeFromDF(projectTheme_df,2)    
code_df.head(8)

code                      name
38     2  Public sector governance
39     2  Public sector governance
40     2  Public sector governance
41     2  Public sector governance
42     2  Public sector governance
43     2  Public sector governance
44     2  Public sector governance
45     2  Public sector governance

In [42]:
projectTheme_df = correct_missing(projectTheme_df,2,'Public sector governance')

In [43]:
code_df = printSelectedCodeFromDF(projectTheme_df,7)    
code_df.head(6)

code                         name
643     7  Social dev/gender/inclusion
644     7  Social dev/gender/inclusion
645     7  Social dev/gender/inclusion
646     7  Social dev/gender/inclusion
647     7  Social dev/gender/inclusion
648     7  Social dev/gender/inclusion

In [44]:
projectTheme_df = correct_missing(projectTheme_df,7,'Social dev/gender/inclusion')

In [45]:
code_df = printSelectedCodeFromDF(projectTheme_df,4)    
code_df.head(8)

code                                      name
252     4  Financial and private sector development
253     4  Financial and private sector development
254     4  Financial and private sector development
255     4  Financial and private sector development
256     4  Financial and private sector development
257     4  Financial and private sector development
258     4  Financial and private sector development
259     4  Financial and private sector development

In [46]:
projectTheme_df = correct_missing(projectTheme_df,4,'Financial and private sector development')

In [47]:
code_df = printSelectedCodeFromDF(projectTheme_df,10)    
code_df.head(14)

code               name
1033    10  Rural development
1034    10  Rural development
1035    10  Rural development
1036    10  Rural development
1037    10  Rural development
1038    10  Rural development
1039    10  Rural development
1040    10  Rural development
1041    10  Rural development
1042    10  Rural development
1043    10  Rural development
1044    10  Rural development
1045    10  Rural development
1046    10  Rural development

In [48]:
projectTheme_df = correct_missing(projectTheme_df,10,'Rural development')

In [49]:
code_df = printSelectedCodeFromDF(projectTheme_df,9)    
code_df.head(8)

code               name
983     9  Urban development
984     9  Urban development
985     9  Urban development
986     9  Urban development
987     9  Urban development
988     9  Urban development
989     9  Urban development
990     9  Urban development

In [50]:
projectTheme_df = correct_missing(projectTheme_df,9,'Urban development')

In [51]:
code_df = printSelectedCodeFromDF(projectTheme_df,3)    
code_df.head(6)

code         name
237     3  Rule of law
238     3  Rule of law
239     3  Rule of law
240     3  Rule of law
241     3  Rule of law
242     3  Rule of law

In [52]:
projectTheme_df = correct_missing(projectTheme_df,3,'Rule of law')

In [53]:
projectTheme_df.name.value_counts().head(15)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64

In [54]:
projectTheme_df.head(10)

code                 name
0     1  Economic management
1     1  Economic management
2     1  Economic management
3     1  Economic management
4     1  Economic management
5     1  Economic management
6     1  Economic management
7     1  Economic management
8     1  Economic management
9     1  Economic management